### ISSUE TITLE
[R] - Should the returning model object gain a class?

### ISSUE THREAD

**Issue by cgiachalis** [mlpack/mlpack#3722]
### What is the desired addition or change?

R's mlpack models should have class attribute in order to build methods on top.

### What is the motivation for this feature?

Currently an mlpack model is a list object. It should be (ideally) an object of class `mlpack` with subclass `<modelType>`. If someone wants to know what the model is, he has to extract the attribute `type` from `output_model` element. This element holds the model's pointer.

For example, `mlpack::Serialise()` creates a function on the fly [model_serialization_function](https://github.com/cran/mlpack/blob/4c4cf2d93e216c1bc196696a3768da8352accb17/R/serialization.R#L8) that serves as an implicit dispatch method for native serialisation routines using `switch` construct. This could have been done easily and more robust using an `S3` method with a class mlpack or `<modelType>`.

``` r
set.seed(1234)
x <- matrix(rnorm(10*5), ncol = 5)

res <- mlpack::knn(query = x, reference = x, k = 3)

class(res)
# [1] "list"

attributes(res$output_model)
$type
# [1] "KNNModel"

```
### If applicable, describe how this feature would be implemented.

We add a class attribute in order to work with S3 OO system.
```r
# add subclass and class
class(res) <- c("KNNModel", "mlpack")

# check
class(res)
# [1] "KNNModel" "mlpack" 
```
That's it.
As an example we can now create a method `serialise_mlpack` to return the raw vector given the model type.

``` r
# New S3 generic for native serialisation
serialise_mlpack <- function(object, ...) {
  UseMethod("serialise_mlpack")
}

# Optional
serialise_mlpack.default <- function(object, ...) {

  stop(sprintf("No method for object %s. See ?serialise_mlpack for details.",
               sQuote(deparse(substitute(object)))))
}

# method for KNNModel
serialise_mlpack.KNNModel <- function(object, ...) {
mlpack:::SerializeKNNModelPtr(object$output_model)

}
```

``` r
# native raw vector (nice to have it exported!!)
serialise_mlpack(res)
```

*So what is the benefit*

Consider the usage of `mlpack::Serialise()`: We invoke it by passing the `model_ouput` element, e.g `Serialize(res$output_model)`. That's not ideal and neither obvious from the help page.

``` r
# Using serialise_mlpack() method internally  
Serialize <- function(object, filename) {
  model_serialization <- serialise_mlpack(object)

  con <- file(as.character(filename), "wb")
  serialize(model_serialization, con)
  close(con)
}

res2 <- mlpack::knn(query = x, reference = x, k = 3)
Serialize(res2, filename = "Hello.rds")
```

Additionally, we can export `serialise_mlpack()` in order to build custom R objects and then serialise to different methods and send it to a storage engine. That was the impetus for opening this issue. Here is an example:

``` r

super_model <- list(desc = "A very secret model",
                    flavour = "mlpack",
                    model = serialise_mlpack(object),
                    version = packageVersion("mlpack"))

# Alternative to base::serialize()                  
qs::qserialize(super_model)                 
```

Also, by adding a class we help other developers to identify `mlpack` object and incorporate to their packages [rstudio/bundle](https://github.com/rstudio/bundle).

### Additional information?

Upon request.

**Comment by rcurtin** [comment]
I know this one's sat open for quite a while---I'm pretty underwater so it would be hard for me to devote the time soon.  The feature definitely seems useful and it would be great to implement.  If anyone wants to give it a shot, I think it shouldn't be all that hard---since mlpack's R bindings are generated, the changes consist of basically just adding some additional code, probably to the end of `src/mlpack/bindings/R/print_R.cpp`, that simply print the necessary class definition and then serialization function.

I guess that `serialize_mlpack` and `serialize_mlpack.default` would need to be added to some new file `serialization.R` in `src/mlpack/bindings/R/mlpack/R/` and then that file should be added to the list `R_SOURCES` in `src/mlpack/bindings/R/CMakeLists.txt`.

I suppose some simple serialization tests should also be added to `src/mlpack/bindings/R/tests/`.

Then to build mlpack, you can use this to set up a build directory from the root directory of the repository:

```
mkdir build/
cd build/
cmake -DBUILD_R_BINDINGS=ON ../
```

And then building the R bindings should be as simple as either `make R` or `make r` (can't remember which) although maybe you want, e.g., `make -j8 R` to build with 8 cores or whatever is right for your system.  Then the generated R package will be in `build/src/mlpack/bindings/R/` and you can check the `.R` files to ensure they were generated in the right way.

@cgiachalis I wrote all that out so that you can self-serve on this feature if you like, but if not anyone else who stumbles upon this is more that welcome to give it a shot 😄 (CC: @coatless @eddelbuettel in case you have some extra time :)) I expect I won't be able to look into this soon, unfortunately.

**Comment by cgiachalis** [comment]
Thanks for your response @rcurtin!

Indeed, we need to add the class definition towards the end of `src/mlpack/bindings/R/print_R.cpp`. With respect to native serialisation routines, it's better to tackle it separately so I am about to  open a separate GitHub issue because an S3 method will depend on the class definition which is not implemented yet.

I am on Win OS... :) - well so I decided to spin up the devcontainer from coatless' PR (https://github.com/mlpack/mlpack/pull/3637) which went well until start building it with R binding flag; configuration errors signalled that some R components were missing, so I guess the devcontainer wasn't set-up for this scenario.

Anyway, for start I wanted to test it by adding only the `'mlpack'` class:

```c++
cout << "  class(out) <- \"mlpack\" " << endl;

```

But eventually we need the ModelType as a subclass; where and how do I get that from?

```c++
cout << "  class(out) <- c( \"" << ??? << "\", \"mlpack\") " << endl;
```

**Comment by coatless** [comment]
@cgiachalis sorry about that! The `.devcontainer` was initially setup to handle only working directly on the C++ library to reduce size. 

Regarding the output, that's stored inside of `outputOptions`. As we're interested in the serialization options, we want to modify:

https://github.com/mlpack/mlpack/blob/822c499f5d71fd9c401d45e0b9977d2aca02f7a0/src/mlpack/bindings/R/print_R.cpp#L207-L211

This generates:

```r
  # Add ModelType as attribute to the model pointer, if needed.
  output_model <- GetParamKNNModelPtr(p, "output_model", inputModels)
  attr(output_model, "type") <- "KNNModel"
```

https://github.com/cran/mlpack/blob/956bbce2f3f444aa8adff8cd91eab31a3088f0de/R/knn.R#L171

So, I suppose a slight change to the print class right after:

https://github.com/mlpack/mlpack/blob/822c499f5d71fd9c401d45e0b9977d2aca02f7a0/src/mlpack/bindings/R/print_serialize_util.hpp#L64-L65

would be: 

```cpp
  MLPACK_COUT_STREAM << "  class(" << d.name << ") <- c(\""
      << util::StripType(d.cppType) << "\", \"mlpack\")";
```

to generate:

```r
  class(output_model) <- c("KNNModel", "mlpack")
```

**Comment by coatless** [comment]
@cgiachalis did you ever progress further or?

**Comment by cgiachalis** [comment]
@coatless Nope, unfortunately :).

**Comment by coatless** [comment]
@cgiachalis no worries, going to knock it out now.

**Comment by cgiachalis** [comment]
> 
> So, I suppose a slight change to the print class right after:
> 
> [mlpack/src/mlpack/bindings/R/print_serialize_util.hpp](https://github.com/mlpack/mlpack/blob/822c499f5d71fd9c401d45e0b9977d2aca02f7a0/src/mlpack/bindings/R/print_serialize_util.hpp#L64-L65)
> 
> Lines 64 to 65 in [822c499](/mlpack/mlpack/commit/822c499f5d71fd9c401d45e0b9977d2aca02f7a0)
> 
>  MLPACK_COUT_STREAM << "  attr(" << d.name << ", \"type\") <- \"" 
>      << util::StripType(d.cppType) << "\""; 
> would be:
> 
>   MLPACK_COUT_STREAM << "  class(" << d.name << ") <- c(\""
>       << util::StripType(d.cppType) << "\", \"mlpack\")";
> to generate:
> 
>   class(output_model) <- c("KNNModel", "mlpack")

Hmm,  the code change should be done right before:
https://github.com/mlpack/mlpack/blob/a42defb288d7f4c0d256b43088f83eed7cec5482/src/mlpack/bindings/R/print_R.cpp#L231

and not inside the print class utility as we haven't defined the `out` variable yet; Note the aim is not to subclass the `output_model` element of the list  but the list itself; so we need to write `class(out) <- c(ModelType, "mlpack", "list")`; 

I experimented a little bit using the dev container (works perfect!, thanks @coatless) and had partially success; I believe it needs a stand alone print class utility.

**Comment by eddelbuettel** [comment]
Hi @cgiachalis sorry this fell to the side.  I'd be happy to help hashing things out -- but I will also admit that I feel that we are still in a bit of an odd 'spot' here.

In general, mlpack itself does a fair amount of direct input/output including to the console (as for example in cli mode). Whereas R would potentially replicate some this at its end (i.e. instead of `std::cout` from 'underneath' in the C++ layer that is core to mlpack).  While that seems appealing to me with my R hat, it is also unappealing to me because it would make one of the mlpack 'ports' behave differently.  

The 'standalone print class utility' could of course also be called from R into an existing C++ routine. This would unify the two approaches a bit more.

**Comment by cgiachalis** [comment]
Right, I managed  to get it done i.e., printing `class(out) <- c(ModelType, "mlpack", "list")` for serializable models.

But it's better we should do the same for non serializable models too. Thoughts?

Here's how it looks now:
```r
res <- mlpack::knn(query = x, reference = x, k = 3)
class(res)
#  "KNNModel"  "mlpack" "list"

res <- mlpack::emst(x)
class(res)
# "list"

res <- mlpack::kernel_pca(x, kernel = "gaussian")
class(res)
# "list"
```

**Comment by eddelbuettel** [comment]
Yes we could start with a base class `mlpack` or `mlpack_object` and add a simple s3 print method which ... if we're clever and we can inspect the object (can we?) may call the underlying C++ printer.  In other words a bridge to to the existing i/o routines.

**Comment by cgiachalis** [comment]
Basically, I created a replicate of  `print_serialize_util.h`  ->  `print_serialize_util2.h` and modified one of the  template functions to get the desire output so I can use it in `print_R.cpp` ;

Now I need to tweak another template function to capture the not serializable models like `mlpack::kernel_pca`: 
(***print_serialize_util2.h***) 
https://github.com/mlpack/mlpack/blob/a42defb288d7f4c0d256b43088f83eed7cec5482/src/mlpack/bindings/R/print_serialize_util.hpp#L23-L32

Will do tomorrow and open a draft PR if everything goes well.

**Comment by cgiachalis** [comment]
FYI

```c++

/**
 * For non-matrix serializable types we need to print something.
 */
template<typename T>
void PrintSerializeUtil2(
    util::ParamData& d,
    const typename std::enable_if<!arma::is_arma_type<T>::value>::type* = 0,
    const typename std::enable_if<data::HasSerialize<T>::value>::type* = 0)
{
  /**
   * This gives us code like:
   *
   *     class(out) <- c("<ModelType>", "mlpack", "list")
   */
  MLPACK_COUT_STREAM << "  class(out) <- c(\""
      << util::StripType(d.cppType) << "\", \"mlpack\", \"list\")";
  MLPACK_COUT_STREAM << std::endl;
}

```
Add  the following:
```c++

    // Add ModelType as class to the output.
  cout << "  # Add ModelType as class to the output, if needed."
      << endl;
  for (size_t i = 0; i < outputOptions.size(); ++i)
  {
    util::ParamData& d = parameters.at(outputOptions[i]);
    params.functionMap[d.tname]["PrintSerializeUtil2"](d, NULL, NULL);
  }
```

Just before:
https://github.com/mlpack/mlpack/blob/a42defb288d7f4c0d256b43088f83eed7cec5482/src/mlpack/bindings/R/print_R.cpp#L231

**Comment by eddelbuettel** [comment]
Hm. But why no S3 dispatch when your very Subject: to this issue suggests to use a class (and hence R dispatch) ?

**Comment by cgiachalis** [comment]
> Hm. But why no S3 dispatch when your very Subject: to this issue suggests to use a class (and hence R dispatch) ?

Let's recap to ensure we are synced. 

The R bindings are generated so we need to add programmatically a S3 class definition in `src/mlpack/bindings/R/print_R.cpp`. But we need the model type; per @coatless advice, the model type is stored inside `outputOptions`. For that reason we replicate `PrintSerializeUtil.h`, then modifying it slightly and use it inside `print_R.cpp`. 

It compiles without errors and builds the R tarball; a quick inspection for `mlpack::random_forest`:

```r
  # Add ModelType as attribute to the model pointer, if needed.
  output_model <- GetParamRandomForestModelPtr(p, "output_model", inputModels)
  attr(output_model, "type") <- "RandomForestModel"

  # Extract the results in order.
  out <- list(
      "output_model" = output_model,
      "predictions" = GetParamURow(p, "predictions"),
      "probabilities" = GetParamMat(p, "probabilities")
  )

  # Add ModelType as class to the output, if needed.
  class(out) <- c("RandomForestModel", "mlpack", "list")

  return(out)
```

The above solution works for serializable models, e,g., `knn`, `random_forest`; for not serializable models e,g., `emst` the returning object is a bare list.  

For not serializable models, it seems there is no model type ( via `cppType`). Should we class only the serializable models then?

Let me know if I misunderstood something.

**Comment by cgiachalis** [comment]
Here's the branch https://github.com/cgiachalis/mlpack/tree/3722-r-add-class

**Comment by cgiachalis** [comment]
Perhaps we could use the binding name instead so as all outputs to be classed consistently in the form of `c(binding_name,  "mlpack", "list")`...

See https://github.com/cgiachalis/mlpack/tree/3722-r-add-binding-name-as-class

That gives us:
```r
res <- mlpack::knn(query = x, reference = x, k = 3)
class(res)
#  "knn"  "mlpack" "list"

res <- mlpack::emst(x)
class(res)
#  "emst"  "mlpack" "list"

res <- mlpack::kernel_pca(x, kernel = "gaussian")
class(res)
#  " kernel_pca"  "mlpack" "list"
```

**Comment by rcurtin** [comment]
@cgiachalis Sorry it took me a while to get back to this.

> For not serializable models, it seems there is no model type ( via cppType). Should we class only the serializable models then?

Some of the mlpack bindings, like emst (and some others), don't compute models but instead just results.  That emst binding just computes the minimum spanning tree of a dataset and returns it as an edge list.  Any of the bindings that generate models can then generally reuse those models in subsequent calls... so you could e.g. call `random_forest()`, get a model, then use it again.

To me it would only make sense to have a special R class for the serializable mlpack models---I don't think it's worth worrying about any other case, because the only other things a binding could return are already R objects: int/string/double/matrices.

If you have a set of changes ready that make serializable objects into R classes, I would happily review; but as always, you and Dirk and James are the R experts... so I can only really review it from the C++ side and would be trusting that the generated R code actually does the thing it's supposed to do correctly. :)

**Comment by eddelbuettel** [comment]
The S3 class system in R is extremely lightweight -- just the class attribute as shown above -- and the basis of a simple dispatch mechanism. Things like fitted objects get their own summaries, extractors etc  R is OO in its own way and uses that extensively: there is so much more than "int/string/double/matrices", (And there are two more class systems because ... features.)   So to me, and  _from the R side_ it would make total sense to layer this in here.  But _from the mlpack side_ it may make less sense because R and mlpack would then 'act' slightly differently from other mlpack interfaces.

I am rather torn about this question.  R excels as a 'shell' around all things data because it already does input (from just about anything), output (ie conversion / storage to just about anything) plus dispatching -- as the `mlpack` R package does to the C++ API -- and much more (viz, plots, ...).  But mlpack must be self-sufficient for the other uses besides R so a rather large amount of code and effort on the mlpack side already goes to something we kinda already have over here.  So a big "I do not know and it is hard" from me.

**Comment by rcurtin** [comment]
I don't see how the R bindings for mlpack would be 'different' from this perspective: you would still receive back some kind of model object from the bindings, which you could then pass in to subsequent mlpack bindings calls.  Did I miss something?  To my understanding we are specifically and *only* talking about adding R classes to serializable mlpack model objects, which are returned from some but not all mlpack bindings.

**Comment by eddelbuettel** [comment]
Fair point.  We could experiment with the particular case of serialization as proposed by @cgiachalis and see where that leads us.

**Comment by cgiachalis** [comment]
>  Sorry it took me a while to get back to this.

@rcurtin No worries!

> Did I miss something? To my understanding we are specifically and only talking about adding R classes to serializable mlpack model objects, which are returned from some but not all mlpack bindings.

Initially, yes and as an example I mentioned a `serialise_mlpack` method for serializable mlpack model objects. Then I worked on James's dev container , added subclasses to  serializable mlpack model objects ...and then I realised that we need classes to non  serializable models too. 

This how it looks the initial draft:

```r
res <- mlpack::knn(query = x, reference = x, k = 3)
class(res)
#  "KNNModel"  "mlpack" "list"

res <- mlpack::emst(x)
class(res)
# "list"

res <- mlpack::kernel_pca(x, kernel = "gaussian")
class(res)
# "list"

```

But the above solution doesn't seem right. Say someone wants to create a simple print method, i.e. `This is your mlpack output`. We can do it for `knn` but not for `emst` unless we manually added it; a more convincing example is when you need a method to store outputs from different flavours, so we need classes.

And as @eddelbuettel  mentioned  *The S3 class system in R is extremely lightweight*; basically it's not even a concern because R mlpack binding outputs everything as `list`, so we're adding two subclasses below `list`, that's it.  And  for C++ perspective is much simpler than my first draft, now it is in one liner in `print_R.cpp`

```c++
  // Add binding name as class to the output.
  cout << "  # Add binding name as class to the output." << endl;
  cout << "  class(out) <- c(\"" << bindingName << "\", \"mlpack\", \"list\")" << endl;
```

https://github.com/cgiachalis/mlpack/tree/3722-r-add-binding-name-as-class

So this is what we get:
```r
res <- mlpack::knn(query = x, reference = x, k = 3)
class(res)
#  "knn"  "mlpack" "list"

res <- mlpack::emst(x)
class(res)
#  "emst"  "mlpack" "list"

res <- mlpack::kernel_pca(x, kernel = "gaussian")
class(res)
#  " kernel_pca"  "mlpack" "list"

```

**Comment by cgiachalis** [comment]
Done in #4045
